// @sap/di.code-validation.core/src/lib/validationConfig.js
/*
 Structure of configuration for validation framework:
 {
     "validators" : {
         "dummyValidator1" : {
             "extensions" : ['.js', '.xsjs'],
             "filters": {
                 "levels" :["error"]
             }
         }
     }
 }
 */
var defaultLevels = ["error", "warning", "info"];

// Constructor
function ValidationConfig(configJson) {
    this.config = configJson || {};
    this.config.validators = this.config.validators || {};
    this.config.ignoredPaths = this.config.ignoredPaths || [];
}

// class methods
ValidationConfig.prototype.getValidators = function() {
    return Object.keys(this.config.validators);
};

ValidationConfig.prototype.getExtensions = function(validatorName) {
    if (!this.config.validators[validatorName]) {
        throw new Error(validatorName + " is not in configuration");
    }
    return this.config.validators[validatorName].extensions || [];
};

ValidationConfig.prototype.getIgnoredPaths = function() {
    return this.config.ignoredPaths;
};

ValidationConfig.prototype.getLevels = function(validatorName) {
    if (!this.config.validators[validatorName]) {
        throw new Error(validatorName + " is not in configuration");
    }
    if (!this.config.validators[validatorName].filters || !this.config.validators[validatorName].filters.levels) {
        return defaultLevels;
    }
    return this.config.validators[validatorName].filters.levels;
};

module.exports = ValidationConfig;
