// @sap/di.code-validation.core/grunt_configs/jsonValidator.conf.js
/**
 * Created by I028014 on 23/11/2016.
 */
module.exports = {
    voter: {
        mochaTest: {
            test: {
                options: {
                    reporter: 'mocha-junit-reporter',
                    reporterOptions: {
                        mochaFile: 'tests/jsonValidator/TEST-index.xml'
                    },
                    colors: true,
                    summery: true,
                    captureFile: 'tests/jsonValidator/output.txt', // Optionally capture the reporter output to a file
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: false // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['jsonvalidator/test/**/*Spec.js']
            },
            src: ['jsonvalidator/test/*Spec.js']
        },
        coverage: {
            src: 'jsonvalidator/test/*Spec.js',
            colors: true,
            options: {
                timeout: 20000,
                print: 'summary',
                growlOnSuccess: true,
                excludes: ['jsonvalidator/utils/json_parse.js'],
                check: {
                    branches:90,
                    lines: 90,
                    statements: 90,
                    functions: 90
                },
                coverageFolder: "coverage/jsonValidator",
                reportFormats: ['cobertura', 'html']
            }
        }
    }
}
