// @sap/di.code-validation.core/grunt_configs/jsValidator.conf.js
module.exports = {
    voter: {
        mochaTest: {
            test: {
                options: {
                    reporter: 'mocha-junit-reporter',
                    reporterOptions: {
                        mochaFile: 'tests/jsValidator/TEST-index.xml'
                    },
                    colors: true,
                    summery: true,
                    captureFile: 'tests/jsValidator/output.txt', // Optionally capture the reporter output to a file
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: false // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['jsvalidator/test/**/*Spec.js']
            }
        },

        coverage: {
            src: 'jsvalidator/test/*Spec.js',
            colors: true,
            options: {
                timeout: 20000,
                print: 'summary',
                growlOnSuccess: true,
                //excludes: ['plugins/*'],
                check: {
                    branches: 63,
                    lines: 89,
                    statements: 89,
                    functions: 90
                },
                coverageFolder: "coverage/jsValidator",
                reportFormats: ['cobertura', 'html']
            }
        }
    }

};
